### PR TITLE
Not allowing bound variables under uninterpreted functions with bv-to-int

### PR DESCRIPTION
**Pull request by yoni206**
Solves #9071 .

The current implementation of int-blasting does not support bound variables under uninterpreted functions. The reason is that these function applications must be guarded to be in the right range, but this range-constraint should be under the appropriate quantifier, and not in the top level formula.

This PR throws an error when this case is reached.

### CODE REVIEW

File: src/theory/bv/int_blaster.cpp [hunk 1]
```diff
@@ -545,6 +545,18 @@ Node IntBlaster::translateWithChildren(
     }
     case kind::APPLY_UF:
     {
+      // The preprocessing pass does not support function applications
+      // with bound variables.
+      for (Node child : original)
+      {
+        if (child.getKind() == Kind::BOUND_VARIABLE)
```
Comment (ajreynol) [inline src/theory/bv/int_blaster.cpp:552]: Should this be generalized to `expr::hasBoundVar(child)`, e.g. we don't handle any translation where the children of an APPLY_UF are non-ground?

File: src/theory/bv/int_blaster.cpp [hunk 2]
```diff
@@ -545,6 +545,18 @@ Node IntBlaster::translateWithChildren(
     }
     case kind::APPLY_UF:
     {
+      // The preprocessing pass does not support function applications
+      // with bound variables.
+      for (Node child : original)
```
Comment (ajreynol) [inline src/theory/bv/int_blaster.cpp:550]:
```suggestion
      for (const Node& child : original)
```

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/bv/int_blaster.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/bv/int_blaster.cpp:552]: hasBoundVar is cheap, it is cached globally per node via an attribute.  

In terms of ensuring we catch all errors, I'm not exactly sure what is allowed of bound variables, are all other cases of bound variables handled on subterms? Can you maybe try an example which replaces the problematic bound variable with a composite term (not under another UF)?

Comment (yoni206) [reply src/theory/bv/int_blaster.cpp:552]: About `hasBoundVar`: good to know! thanks.

About sub-terms, you are totally right, of course. I'll fix and add a corresponding regression.

Comment (yoni206) [reply src/theory/bv/int_blaster.cpp]: fixed and pushed.

**Review by ajreynol — APPROVED** [review]

**Pull request merged**